[FLINK-21589][docs] Document table pipeline upgrades (#17260)
apache/flink: docs/content/docs/dev/table/concepts/overview.md

The following pages explain concepts, practical limitations, and stream-specific configuration parameters of Flink's relational APIs on streaming data.

State Management
----------------

Table programs that run in streaming mode leverage all capabilities of Flink as a stateful stream
processor.

In particular, a table program can be configured with a [state backend]({{< ref "docs/ops/state/state_backends" >}})
and various [checkpointing options]({{< ref "docs/dev/datastream/fault-tolerance/checkpointing" >}})
for handling large amounts of state and fault tolerance. It is possible to take a savepoint of a running

> Table API & SQL pipeline and to restore the application's state at later point in time.

Reviewer: ```suggestion
Table API & SQL pipeline and to restore the application's state at a later point in time.
```

### State Usage

> Due to the declarative nature of Table API & SQL program, it is not always obvious where and how much
> state is used within a table pipeline. The planner decides about when state is necessary to compute a correct

Reviewer 2: ```suggestion
Due to the declarative nature of Table API & SQL programs, it is not always obvious where and how much state is used within a table API pipeline. The planner decides when state is necessary to compute a correct
```

result. A pipeline is optimized to claim as little state as possible given the current set of optimizer
rules.

{{< hint info >}}
Source tables are never kept entirely in state. This depends on the used operations.
{{< /hint >}}

> Simple `SELECT ... FROM ... WHERE` queries that only consist of field projections or filters are usually
> stateless pipelines. However, operations such as joins, aggregations, or deduplications require to keep
> intermediate results in a fault tolerant storage for which Flink's state abstractions are used.

Reviewer 2: ```suggestion
Queries such as `SELECT ... FROM ... WHERE` that only consist of field projections or filters are usually
stateless pipelines. However, operations such as joins, aggregations, or deduplications require keeping
intermediate results in a fault-tolerant storage for which Flink's state abstractions are used.
```

{{< hint info >}}

> Please refer to the individual operator documentation for more details about how much state is required
> and how to limit a potentially ever growing state size.

Reviewer 2: ```suggestion
Please refer to the individual operator documentation for more details about how much state is required and how to limit a potentially ever-growing state size.
```

{{< /hint >}}

For example, a regular SQL join of two tables requires the operator to keep both input tables in state
entirely. For correct SQL semantics, the runtime needs to assume that a matching could occur at any
point in time from both sides. Flink provides [optimized window and interval joins]({{< ref "docs/dev/table/sql/queries/joins" >}})
that aim to keep the state size small by exploiting the concept of [watermarks]({{< ref "docs/dev/table/concepts/time_attributes" >}}).

### Stateful Upgrades and Evolution

Table programs that are executed in streaming mode are intended as *standing queries* that statically

> define an end-to-end pipline.

Reviewer: ```suggestion
define an end-to-end pipeline.
```

In case of stateful pipelines, any change to both the query or Flink's planner might lead to a completely
different execution plan. This makes stateful upgrades and the evolution of table programs challenging
at the moment. The community is working on improving those shortcomings.

For example, by adding a filter predicate, the optimizer might decide to reorder joins or change the
schema of an intermediate operator. This prevents restoring from a savepoint due to either changed
topology or different column layout within the state of an operator.

The query implementer must ensure that the optimized plans before and after the change are compatible.
Use the `EXPLAIN` command in SQL or `table.explain()` in Table API to [get insights]({{< ref "docs/dev/table/common" >}}#explaining-a-table).

> Since new optimizer rules are continously added, and operators become more efficient and specialized,

Reviewer: ```suggestion
Since new optimizer rules are continuously added, and operators become more efficient and specialized,
```

The following pages explain concepts, practical limitations, and stream-specific configuration parameters of Flink's relational APIs on streaming data.

State Management
----------------

Table programs that run in streaming mode leverage all capabilities of Flink as a stateful stream
processor.

In particular, a table program can be configured with a [state backend]({{< ref "docs/ops/state/state_backends" >}})
and various [checkpointing options]({{< ref "docs/dev/datastream/fault-tolerance/checkpointing" >}})
for handling different requirements regarding state size and fault tolerance. It is possible to take
a savepoint of a running Table API & SQL pipeline and to restore the application's state at a later
point in time.

### State Usage

> Due to the declarative nature of Table API & SQL program, it is not always obvious where and how much
> state is used within a pipeline. The planner decides whether state is necessary to compute a correct
> result. A pipeline is optimized to claim as little state as possible given the current set of optimizer
> rules.

Reviewer 2: ```suggestion
Due to the declarative nature of Table API & SQL programs, it is not always obvious where and how much
state is used within a pipeline. The planner decides whether state is necessary to compute a correct
result. A pipeline is optimized to claim as little state as possible given the current set of optimizer
rules.
```

{{< hint info >}}
Conceptually, source tables are never kept entirely in state. An implementer deals with logical tables
(i.e. [dynamic tables]({{< ref "docs/dev/table/concepts/dynamic_tables" >}})). Their state requirements

> depend on the used operations.

Reviewer 2: I feel like more context could be provided in this block to make one idea lead to the next.  Wdyt?

Author: I'm not sure if I understand this correctly. Isn't it leading to the next paragraph already?

Reviewer 2: What about the initial

“We don’t keep tables in state but operators. But those might have to store their entire input tables.”

{{< /hint >}}

> Queries such as `SELECT ... FROM ... WHERE` queries that only consist of field projections or filters are usually
> stateless pipelines. However, operations such as joins, aggregations, or deduplications require to keep
> intermediate results in a fault tolerant storage for which Flink's state abstractions are used.

Reviewer 2: ```suggestion
Queries such as `SELECT ... FROM ... WHERE` which only consist of field projections or filters are usually
stateless pipelines. However, operations such as joins, aggregations, or deduplications require keeping
intermediate results in a fault-tolerant storage for which Flink's state abstractions are used.
```

Author: sorry about this, I forgot to commit these changes

{{< hint info >}}
Please refer to the individual operator documentation for more details about how much state is required
and how to limit a potentially ever growing state size.
{{< /hint >}}

For example, a regular SQL join of two tables requires the operator to keep both input tables in state
entirely. For correct SQL semantics, the runtime needs to assume that a matching could occur at any
point in time from both sides. Flink provides [optimized window and interval joins]({{< ref "docs/dev/table/sql/queries/joins" >}})
that aim to keep the state size small by exploiting the concept of [watermarks]({{< ref "docs/dev/table/concepts/time_attributes" >}}).

### Stateful Upgrades and Evolution

> Table programs that are executed in streaming mode are intended as *standing queries* that statically
> define an end-to-end pipeline.

Reviewer 2: ```suggestion
Table programs that are executed in streaming mode are intended as *standing queries* (defined once and then executed continuously) that statically define an end-to-end pipeline.
```
